fix(core): do not trigger CSP alert/report in Firefox and Chrome

peruukki opened the pull request:
Default to using DOMParser if it is available and fall back to createDocument if needed. This is the approach used by DOMPurify and suggested in the related Angular.js pull request angular/angular.js#17013. It also safely avoids using an inline style tag that causes CSP violation errors if inline CSS is prohibited.

The related unit tests in `html_sanitizer_spec.ts`, "should not allow JavaScript execution when creating inert document" and "should not allow JavaScript hidden in badly formed HTML to get through sanitization (Firefox bug)", are left untouched to assert that the behavior hasn't changed in those scenarios.

Fixes #25214.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

If [innerHTML] is used in a component and the Content-Security-Policy does not allow inline styles, a CSP violation will be reported by at least Firefox and Chrome.

Issue Number: #25214 

## What is the new behavior?

No CSP violation is reported.

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

## Other information

Like mentioned above, there is an existing unit test related to this change that tests that badly-formatted HTML is sanitized properly, and I didn't change the test because the behavior shouldn't have changed on that regard. I don't think there is a convenient way to test CSP violations, so I didn't write any new tests.

This is my first contribution attempt for Angular. 😃

peruukki committed:
fix(core): do not trigger CSP alert/report in Firefox and Chrome (#36578)

If [innerHTML] is used in a component and a Content-Security-Policy is set
that does not allow inline styles then Firefox and Chrome show the following
message:

> Content Security Policy: The page’s settings observed the loading of a
resource at self (“default-src”). A CSP report is being sent.

This message is caused because Angular is creating an inline style tag to
test for a browser bug that we use to decide what sanitization strategy to
use, which causes CSP violation errors if inline CSS is prohibited.

This test is no longer necessary, since the `DOMParser` is now safe to use
and the `style` based check is redundant.

In this fix, we default to using `DOMParser` if it is available and fall back
to `createHTMLDocument()` if needed. This is the approach used by DOMPurify
too.

The related unit tests in `html_sanitizer_spec.ts`, "should not allow
JavaScript execution when creating inert document" and "should not allow
JavaScript hidden in badly formed HTML to get through sanitization (Firefox
bug)", are left untouched to assert that the behavior hasn't changed in
those scenarios.

Fixes #25214.

peruukki commented:
@googlebot I signed it!

peruukki commented:
The lint failures are in the file I changed but not in my changes, not sure if I should fix them.

koto commented:
As explained in https://github.com/angular/angular.js/pull/17013#issuecomment-614603071, please don't change `style` to `span`. Style is crucial for the payload there.

peruukki commented:
Thanks for the info and help, I'll try to get the fix right in Angular.js first and then come back to update this.

peruukki commented:
I haven't heard back from the Angular.js pull request but updated this one anyway with a similar fix.

peruukki commented:
Also, like I mentioned in https://github.com/angular/angular.js/pull/17013#issuecomment-618524832, we might want to make the DOMParser existence check stricter and verify that the HTML support version of `parseFromString` works, if the check is run by default on all browsers.

peruukki committed:
refactor(core): split inert strategies to separate classes (#36578)

The `inertDocument` member is only needed when using the InertDocument
strategy. By separating the DOMParser and InertDocument strategies into
separate classes, we can easily avoid creating the inert document
unnecessarily when using DOMParser.

peruukki committed:
fix(core): determine required DOMParser feature availability (#36578)

Verify that HTML parsing is supported in addition to DOMParser existence.
This maybe wasn't as important before when DOMParser was used just as a
fallback on Firefox, but now that DOMParser is the default choice, we need
to be more accurate.

peruukki commented:
I refactored the two strategies to separate classes and made the DOMParser availability check more accurate.

Splaktar commented:
[CircleCI](https://circleci.com/gh/angular/angular/691520) is indicating that this change makes the `main-2015` bundle 500 bytes smaller.

It wants you to update the following file and push up those changes for that test to pass:
```bash
FAIL: Commit e48607b1e0767d70d51945b5a787d8362a5f26a7 uncompressed main-es2015 fell below expected size by 500 bytes or >1% (expected: 452060, actual: 451537).
If this is a desired change, please update the size limits in file '/home/circleci/ng/goldens/size-tracking/aio-payloads.json'.
```

Splaktar reviewed (approved):
LGTM

peruukki committed:
ci: decrease expected AIO and integration payload sizes (#36578)

The changes in #36687 removed enough code to exceed the CI check limits, so
the expected main-es2015 size needs adjusting.

IgorMinar reviewed (approved):
Thanks for this great cleanup @peruukki!

This change LGTM, but it will need to be rebased before it can be merged. 

I see that @koto commented on this PR as well. Is it because this change was initiated by @koto in some way or was it just a drive-by comment?

I would love for @petebacondarwin to also take a look at this change as he was involved in changing the sanitization the most recently and should consistently apply this cleanup to AngularJS as well if it's possible to do so there.

petebacondarwin reviewed (approved):
LGTM!

I merge the equivalent fix in AngularJS.

Can you change the `fix` commit message to be more oriented towards the thing being fixed rather than the internal refactoring? I did this unilaterally in the AngularJS commit: https://github.com/angular/angular.js/commit/2fab3d4e00f4fe35bfa3cf255160cb97404baf24

Also there are conflicts that need resolving via a rebase.

koto commented:
LGTM, thanks!

peruukki commented:
> Can you change the `fix` commit message to be more oriented towards the thing being fixed rather than the internal refactoring? I did this unilaterally in the AngularJS commit: [angular/angular.js@2fab3d4](https://github.com/angular/angular.js/commit/2fab3d4e00f4fe35bfa3cf255160cb97404baf24)
>
> Also there are conflicts that need resolving via a rebase.

Oh yeah, that's a lot more useful message, I changed the PR titles accordingly too. I also largely copied your commit message to this one, thanks!

The conflicts should be resolved now, I've rebased a couple times already but they keep coming back. 😄

petebacondarwin commented:
Sadly the conflicts have come back again!

peruukki commented:
> Sadly the conflicts have come back again!

This is turning out to be the hardest part. 😅 Conflict-free again, for now.

peruukki commented:
Now I get this error:
> FAIL: Commit undefined uncompressed polyfills-es2015 fell below expected size by 500 bytes or >1% (expected: 37640, actual: 37248).
> If this is a desired change, please update the size limits in file '/home/circleci/.cache/bazel/_bazel_circleci/9ce5c2144ecf75d11717c0aa41e45a8d/execroot/angular/bazel-out/k8-fastbuild/bin/integration/cli-hello-world-ivy-i18n_test.sh.runfiles/angular/goldens/size-tracking/integration-payloads.json'.

But I wouldn't expect my changes to affect the polyfills bundle, and looks like the same error comes in the master branch too. So maybe I'll wait for its resolution in the master branch and try again later?

petebacondarwin commented:
Yes. Sorry. Looks like its something we need to fix.

AndrewKushnir commented:
Hi @peruukki, thanks for rebasing your PR. There was a problem in master (the payload size check was failing), which is now resolved and your PR should go "green" after rebase (and resolving a conflict in `goldens/size-tracking/integration-payloads.json`). Thank you.

peruukki commented:
Thanks for the update @AndrewKushnir. I rebased and now the tests pass, no conflicts either at the moment.

AndrewKushnir commented:
Thanks for rebasing this PR @peruukki 👍

I've started [blueprint](http://test/OCL:316305247:BASE:316305249:1592106243366:e322c84b) and [global](https://test.corp.google.com/OCL:316305247:BASE:316305346:1592106365764:b35b438) presubmits in Google's codebase and will update this thread with results as soon as I have them.

Thank you.

IgorMinar reviewed (approved):
One nit about code duplication in tests, but otherwise LGTM! Can you pleas make that change as a fixup commit? (once you do we should not need presubmits to reverify the change so we can speed up the merging process.

Thank you for this great cleanup!

Reviewed-for: global-approvers

AndrewKushnir commented:
FYI, all Google codebase checks are successful, so I removed the `presubmit` label. I've left the "cleanup" label though, since there is an additional feedback from Igor. Thank you.

peruukki commented:
> One nit about code duplication in tests, but otherwise LGTM! Can you pleas make that change as a fixup commit? (once you do we should not need presubmits to reverify the change so we can speed up the merging process.
> 
> Thank you for this great cleanup!
> 
> Reviewed-for: global-approvers

Thanks @IgorMinar for the improvement suggestion, I made the change but I'm not sure I did it like you intended, it's still pending something related to presubmit. Should I have just created a fixup commit but not squash it?

Splaktar commented:
> Thanks @IgorMinar for the improvement suggestion, I made the change but I'm not sure I did it like you intended, it's still pending something related to presubmit. Should I have just created a fixup commit but not squash it?

Yes, it needed to stay a separate, unsquashed commit in this case.

IgorMinar commented:
lgtm now! thank you very much @peruukki!

We are currently in a code freeze mode so no PRs are being merged until v10.0.0 is cut later this week. I expect this PR to be merged as soon as the code freeze is lifted after the release.

AndrewKushnir commented:
Hi @@peruukki, FYI I rebased your PR on top of the latest master to make sure the payload size limits are still in the right range (since I merged a few more PRs). Once CI goes "green", this PR will be a part of the merge queue. Thank you.

IgorMinar commented:
I cherry-picked this change in the 10.0.x branch as https://github.com/angular/angular/pull/37783
